sdk/provisioning/Azure.Provisioning/src/Infrastructure.cs

```diff
@@ -93,6 +94,97 @@ public virtual void Remove(Provisionable resource)
+    /// <returns>Whether the name is a valid bicep identifier name.</returns>
+    public static bool IsValidIdentifierName(string? identifierName)
+    {
+        if (string.IsNullOrEmpty(identifierName)) { return false; }
+        if (char.IsDigit(identifierName![0])) { return false; }
+        foreach (char ch in identifierName)
+        {
+            if (!char.IsLetterOrDigit(ch) && ch != '_')
+            {
+                return false;
+            }
+        }
+        return true;
+    }
+
+    /// <summary>
+    /// Validates whether a given bicep identifier name is correctly formed of
+    /// letters, numbers, and underscores.
+    /// </summary>
+    /// <param name="identifierName">The proposed bicep identifier name.</param>
+    /// <exception cref="ArgumentNullException">Throws if null.</exception>
+    /// <exception cref="ArgumentException">Throws if empty or invalid.</exception>
+    public static void ValidateIdentifierName(string? identifierName)
+    {
+        if (identifierName is null)
+        {
+            throw new ArgumentNullException(nameof(identifierName), $"{nameof(identifierName)} cannot be null.");
+        }
+        else if (identifierName.Length == 0)
+        {
+            throw new ArgumentException($"{nameof(identifierName)} cannot be empty.", nameof(identifierName));
+        }
+        else if (char.IsDigit(identifierName[0]))
+        {
+            throw new ArgumentException($"{nameof(identifierName)} cannot start with a number: \"{identifierName}\"", nameof(identifierName));
+        }
+
+        foreach (var ch in identifierName)
+        {
+            if (!char.IsLetterOrDigit(ch) && ch != '_')
+            {
+                throw new ArgumentException($"{nameof(identifierName)} should only contain letters, numbers, and underscores: \"{identifierName}\"", nameof(identifierName));
+            }
+        }
+    }
+
+    /// <summary>
+    /// Normalizes a proposed bicep identifier name.  Any invalid characters
+    /// will be replaced with underscores.
+    /// </summary>
+    /// <param name="identifierName">The proposed bicep identifier name.</param>
+    /// <returns>A valid bicep identifier name.</returns>
+    /// <exception cref="ArgumentNullException">Throws if null.</exception>
+    /// <exception cref="ArgumentException">Throws if empty.</exception>
+    public static string NormalizeIdentifierName(string? identifierName)
+    {
+        if (identifierName is null)
+        {
+            // TODO: This may be relaxed in the future to generate an automatic
+            // name rather than throwing
+            throw new ArgumentNullException(nameof(identifierName), $"{nameof(identifierName)} cannot be null.");
+        }
+        else if (identifierName.Length == 0)
+        {
+            throw new ArgumentException($"{nameof(identifierName)} cannot be empty.", nameof(identifierName));
+        }
+
+        StringBuilder builder = new(identifierName.Length);
```

Comment:
Probably doesn't matter too much, but if `identifierName` is already valid, why allocate a StringBuilder and a new string?

Reply:
Fixed